### ISSUE TITLE
Ensure webpack meets NodeJS ESM Spec (Or prepare for it). 

### ISSUE THREAD

**Issue by TheLarkInn** [webpack/webpack#4170]
After a discussion I discovered, @ljharb had [expressed some troublesome concerns](https://github.com/airbnb/babel-plugin-dynamic-import-webpack/pull/14#issuecomment-275159815) about webpack 2. Whether they are valid yet or not I haven't confirmed based on spec (cc @Kovensky), but I think we should work together with any TC-39'rs on the node side if there are descrepencies we are missing. In the end we aim to be as close to spec as possible.

**Comment by ljharb** [comment]
For the record, I don't think the ES spec is where the conflict arises - the ES spec doesn't contain "modules" in a conceptual sense, it contains "the module parsing goal" and "import/export syntax". The additional required piece for "modules" to exist is "loading semantics" - which, absent agreement on the loader spec, comprises "the intersection of browsers' and node's semantics".

**Comment by sokra** [comment]
Let me quote the nodejs esm spec before discussing:

https://github.com/nodejs/node-eps/blob/master/002-es6-modules.md#54-es-consuming-commonjs

### 5.4. ES consuming CommonJS

#### 5.4.1. default imports

`module.exports` is a single value. As such it does not have the dictionary
like properties of ES module exports. In order to facilitate named imports for
ES modules, all properties of `module.exports` will be hoisted to named exports
after evaluation of CJS modules with the exception of `default` which will
point to `module.exports` directly.

##### 5.4.1.1. Examples

Given:

```javascript
// cjs.js
module.exports = {
  default:'my-default',
  thing:'stuff'
};
```

You will grab `module.exports` when performing an ES import.

```javascript
// es.js

// grabs the namespace
import * as baz from './cjs.js';
// baz = {
//   get default() {return module.exports;},
//   get thing() {return this.default.thing}.bind(baz)
// }

// grabs "default", aka module.exports directly
import foo from './cjs.js';
// foo = {default:'my-default', thing:'stuff'};

// grabs "default", aka module.exports directly
import {default as bar} from './cjs.js';
// bar = {default:'my-default', thing:'stuff'};
```

------

Given:

```javascript
// cjs.js
module.exports = null;
```

You will grab `module.exports` when performing an ES import.

```javascript
// es.js
import foo from './cjs.js';
// foo = null;

import * as bar from './cjs.js';
// bar = {default:null};
```

------

Given:

```javascript
// cjs.js
module.exports = function two() {
  return 2;
};
```

You will grab `module.exports` when performing an ES import.

```javascript
// es.js
import foo from './cjs.js';
foo(); // 2

import * as bar from './cjs.js';
bar.name; // 'two'
bar.default(); // 2
bar(); // throws, bar is not a function
```

------

Given:

```javascript
// cjs.js
module.exports = Promise.resolve(3);
```

You will grab `module.exports` when performing an ES import.

```javascript
// es.js
import foo from './cjs.js';
foo.then(console.log); // outputs 3

import * as bar from './cjs.js';
bar.default.then(console.log); // outputs 3
bar.then(console.log); // throws, bar is not a Promise
```

**Comment by ljharb** [comment]
It's worth noting that there have been recent changes which are not yet written up in an EP (which conflict with what you posted); an updated one will hopefully be posted this month.

These things move too fast to be able to implement software based on it and expect it will stay correct for long.

**Comment by TheLarkInn** [comment]
@ljharb Just to confirm: Is this the latest we stand with the specification? https://github.com/nodejs/node-eps/pull/39

**Comment by ljharb** [comment]
It's the latest that's been written up, but not the latest that's been decided.

**Comment by sokra** [comment]
We can take another look when the EP is updated.

Until then we stick with the latest decided spec.

Also keep in mind that webpack tries to support as many specs as possible. I. e. if possible we also want to support the old babel syntax (maybe with deprecation warning), because we don't want to break code.

**Comment by ljharb** [comment]
@sokra there is nothing "decided". The point of the EPs is that it's the *intention*, but in no way is it ratified or certain. (Note that I'm not advocating that you change anything now, so it's fine to wait til the next EP update - I'm advocating that you support *nothing* until there's actually a decided spec - and I realize that that ship has long since sailed for webpack.)

**Comment by TheLarkInn** [comment]
Maybe we could provide some sort of code mod happens for most use cases.

**Comment by ljharb** [comment]
That's a great idea; Babel will need one too.

**Comment by sokra** [comment]
> I'm advocating that you support nothing until there's actually a decided spec

That is not an option. We can't omit ES2015 modules or interop.

**Comment by ljharb** [comment]
@sokra the best option would have been to continue using babel, so that the ecosystem wasn't inflicted with yet *another* incompatible interop.

**Comment by sokra** [comment]
https://twitter.com/wSokra/status/828902492893241344

**Comment by TheLarkInn** [comment]
It is best to put this in the communities hands. I feel we could support interop limit the breaking changes, and use webpack/webpack-cli to help author a code mod, but waiting on a stable spec is important.

**Comment by ljharb** [comment]
Note that I'm not advocating that webpack 2 drop support for ESM. I'm simply explaining my belief, which is that it *should* have stuck with webpack 1/Babel's approach.

At this point, the best thing for users is likely "no additional changes until the spec finalizes".

**Comment by TheLarkInn** [comment]
@ljharb I think thats best. Any thing we should know about the async loading that got agreed to be researched last night?

**Comment by ljharb** [comment]
Nothing yet, since it's just in a research phase.

**Comment by sokra** [comment]
babel approach doesn't allow tree shaking.

**Comment by ljharb** [comment]
@sokra a) the spec doesn't care about tree shaking, b) i'm not sure why it doesn't - almost all npm-installed modules use `require`, so if you can tree-shake those, you can tree-shake anything.

**Comment by mikegreiling** [comment]
maybe it would be prudent to place a warning somewhere in the documentation that the behavior of `import { foo } from 'common-js-module'` may change at some future date so at least people won't be caught off guard

**Comment by davidpelayo** [comment]
What @mikegreiling is the best option I think.

**Comment by graingert** [comment]
@mikegreiling or in the release notes of webpack 3. (because this change would be breaking)

**Comment by evilebottnawi** [comment]
I think we can close it, webpack@4 support ES modules, webpack@5 supports more cases, for ES modules output we have the other issue, but if you found something broken/miesleading/etc, feel free to open a new issue
